WIP: Add route support to RM (async)

wraythex opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This is a work in progress to add Route support to RM async branch.
May not be ready for everyone to test yet.  Please check PR comments for updates and progress info.

## Description
<!--- Describe your changes in detail -->
This pulls route info from the db, sends it to the front end,
adds a marker for the start and end points, adds a polyline for the path,
adds a Routes toggle switch in the Pokestop drop down,
adds an option to disable showing any routes,
and adds the rest of the processing needed for routes.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
New features need new code!

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This is a work in progress.  I'll be testing it on my own server.
I'll update this PR with comments on my progress.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--  NOTE: In order to check code style locally and avoid having your build rejected by Travis, -->
<!--  run the following commands before you commit: `flake8 .` and `npm run lint`. Fix any -->
<!--  issues they point out. Note also that flake's NOQA is disabled on Travis. -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

wraythex commented:
Most of the route functionality is in place.

Even though routes start and end at a particular poi, the path on the map does not always connect to those poi, just like in the game.

I have placed markers next to the start and end of routes.
Blue flag for start, Pink flag for end, just like in the game.

I have placed the markers to right of the poi they are associated with.  I did this so that the markers wouldn't be covered up if you are showing the stop or gym on the map.

I placed the start (blue) marker a little higher than the end (pink) marker.  I did this because some routes have start/end at the same poi and I wanted both to be visible.

There are a couple of issues as of this comment:
1) Turning on the routes toggle switch does not immediately show the routes on the map
 - you have to turn the switch on, and then reload the map to see the routes
2) Sometimes turning the toggle off won't remove all routes from the map
 - reloading the map with the toggle off will remove all routes

If anyone can see what I'm missing with regard to showing/hiding these routes, I'd appreciate the help.

Other than those issues, I think this is ready for some testing by others.

michikrug commented:
Thanks a lot for your work.
I have taken a look at your code and fixed your stated issues.
Will probably create a PR towards your branch so that you can incorporate my changes. :)

wraythex commented:
That would be great!  Thank you very much for the help.  I really appreciate it.

michikrug commented:
Done

michikrug commented:
One issue I just spotted: the routes toggle is currently positioned in the "Pokestop" section.
If we disable Pokestops, the routes still show but can not be toggled anymore as the section is hidden completely.
We maybe should move the routes toggle to the "Misc" section

michikrug commented:
I will add the change to my PR

wraythex commented:
Ah, good catch on spotting that variable I missed.  :)
I do like the placement of the switch in the misc section better.  Thanks for that too.

This does fix the first problem of routes not loading when turning the button on.
However, this still doesn't quite fix the rare issue of routes not unloading when turning the button off.

This second issue is a little harder to encounter.  I've only seen it when I've turned routes on and then left them sitting there for "a while" (maybe 15-60 minutes?).  It's hard to reproduce, so I'm not sure what's going on.

My guess is that either a path and two markers are being created (without the associated pop-up info) but not saved into the mapData to be deleted when turning the switch off.  Or, a path and two markers are created, but are then somehow orphaned by a different update call.  I'll try to investigate more tomorrow.

Thanks again for your help.  I think once we fix this last issue it should be good for everyone.

michikrug commented:
Interesting, followed your idea and had the map open for about 20mins and could observe the same behaviour. No errors in console. Strange.

wraythex commented:
I believe I have fixed the problem.  It was caused by creating new markers and storing them in mapData which orphaned the original markers that were still on the map.

This should be ready for testing by anyone on the async branch now.

JabLuszko commented:
![image](https://github.com/cecpk/RocketMAD/assets/1145742/ee68edb9-b703-4e24-8898-8de348bda363)
Please connect start_poi to first waypoint and last waypoint to end_poi so the route will start/end at poi rather than at first waypoint.

wraythex commented:
I connected the first waypoint to the start poi and the last waypoint to the end poi.  It won't match what is in the game, but should make more sense on the map.

I also unified the popups, they will all show the same info now.

I fixed where the popup will display on routes.  It will popup where the mouse cursor is.  And it will even move with the mouse across different parts of the path.

Luckily I found a post online saying that adding "padding" to the map will fix routes disappearing.  I've added that and it fixed the issue on the routes I can see.

Let me know if all of this works for you.

JabLuszko commented:
Everything is complained about seems to be working, good job :-)

michikrug commented on static/js/map/map.routes.js:
```diff
@@ -0,0 +1,252 @@
+}
+
+// left pad a string/input to a fixed length with extra padString
+function zpad(input, length, padString) {
+  let s = input.toString()
+  while (s.length < length) {
+    s = padString + s
+  }
+  return s
+}
```
I guess you could use the already existing padStart function

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart

wraythex replied:
RocketMAD already had an lpad function, I'll use that.

michikrug commented on static/js/map/map.routes.js:
```diff
@@ -0,0 +1,252 @@
+        iconUrl = `static/images/routes/route_${marker.start ? 'start' : 'end'}.png`
+        routeTitle = `Route ${marker.start ? 'start' : 'end'}`
+    }
+    let rDS = parseInt(route.route_duration_seconds,10)
```
```suggestion
    const rDS = parseInt(route.route_duration_seconds,10)
```

michikrug commented on static/js/map/map.routes.js:
```diff
@@ -0,0 +1,252 @@
+        routeTitle = `Route ${marker.start ? 'start' : 'end'}`
+    }
+    let rDS = parseInt(route.route_duration_seconds,10)
+    let duration = `${Math.floor(rDS/3600)}h ${zpad( Math.floor((rDS%3600)/60), 2, '0')}m ${zpad( Math.floor(rDS%60), 2, '0')}s`
```
```suggestion
    const duration = `${Math.floor(rDS/3600)}h ${zpad( Math.floor((rDS%3600)/60), 2, '0')}m ${zpad( Math.floor(rDS%60), 2, '0')}s`
```

JimBim89 commented:
Thanks for working on this, some small suggestions:

1. Instead of 2x the route icon in the mouse over route tooltip one could be replaced by the "image" data.
2. Add additional data like "description", "tags" or "image_border_color_hex" (around the route "image"?).
3.  Its allowed to have start and stop on the same pokestop, is this correctly displayed?
-> A dual colored icon might be nice here and a merged tooltip that says start and stop.
Not sure if there are differences in "type" or "path_type" for those or if you need to check for start=end.

wraythex commented:
1) I'll update the start/stop/route pop-ups to have the images.  That shouldn't be too difficult.
2) I can look into the image border color, but I think description and tags starts making the pop-up pretty crowded.  I'd recommend leaving that for a future update if it's still wanted.
3) Yes, start/stop is correctly displayed.  I have offset the icons by 8 pixels so that you can see both start/stop at the same poi. (start icon is 8 pixels higher than stop icon)

JimBim89 commented on static/js/map/map.routes.js:
```diff
@@ -0,0 +1,265 @@
+                ${i18n('Last scanned')}: <strong>${timestampToDateTime(route.last_updated)}</strong>
+              </div>
+              <div>
+                <a href='javascript:void(0);' onclick='javascript:openMapDirections(${route.start_poi_latitude},${route.start_poi_longitude},"${settings.mapServiceProvider}");' title='${i18n('Open in')} ${mapServiceProviderNames[settings.mapServiceProvider]}'><i class="fas fa-map-marked-alt"></i> ${route.start_poi_latitude.toFixed(5)}, ${route.start_poi_longitude.toFixed(5)}</a>
```
All tooltips have the start lat long values. This should be adjusted for the 3 different tooltips.

JimBim89 commented:
And the route path tooltip can't be "activated" via a mouse click to stay open.

JimBim89 commented on rocketmad/models.py:
```diff
@@ -975,6 +975,121 @@ def get_nests(swLat, swLng, neLat, neLng, oSwLat=None, oSwLng=None,
+            Routes.route_distance_meters, Routes.route_duration_seconds,
+            Routes.start_poi_latitude, Routes.start_poi_longitude,
+            Routes.end_poi_latitude, Routes.end_poi_longitude,
+            Routes.image_border_color_hex, Routes.image, Routes.tags,
```
Using this to color the poi-image borders Purple, Blue, Green, Yellow or Red would look nice.

wraythex replied:
I will add the "on click" function to the path.

But, I don't know what you're suggesting for coloring "the poi-image borders".  All I see here is a list of 9 variables.

JimBim89 replied:
Use the color value from image_border_color_hex to add a border to the images.
Like the blue borders on pokestops and the team color borders around gym images.

wraythex commented:
Added "on click".  This allows mobile users to show the popups.  Not really beneficial to desktop users since they can hover the mouse where ever they want, and "mouse move" and "mouse out" tend to happen right after an "on click".

Added image_border_color_hex ring around the images.

Added both start and stop lat/lng to all pop-ups.  Making it dynamic for paths would be very difficult and not beneficial since you have to start a route at the start or end.
